storagenode/piecestore/endpoint.go

```diff
@@ -315,6 +338,24 @@ func (endpoint *Endpoint) Upload(stream pb.DRPCPiecestore_UploadStream) (err err
 	defer commitOrderToStore(ctx, &largestOrder)
 
 	for {
+		// Measure upload rate (bytes per second)
+		dt := time.Since(startTime)
+
+		// Skip flagging unusually slow connections in the first moments of uploading.
+		if dt > endpoint.config.DelayUntilSlowConnectionFlagged {
```

Comment:
I would move the whole logic inside the helper, so the code here could be.

```
if endpoint.uploadIsTooSlow(&startTime, pieceWriter.Size()) {
     return rpcstatus.Errorf(rpcstatus.Aborted, "upload rate falls below limit")
}
```

Or with logging, to help diagnostics:

```
if err := endpoint.verifyUploadSpeed(&startTime, pieceWriter.Size()); err != nil {
     endpoint.log.Debug("slow upload cancelled", zap.Error(err))
     return rpcstatus.Errorf(rpcstatus.Aborted, "upload rate falls below limit")
}
```

Reply:
done